target/riscv/csr.c

```diff
@@ -1347,19 +1343,38 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
     // The capability cause has moved to xTVAL so we don't report it here.
     RISCVCPU *cpu = env_archcpu(env);
     target_ulong ccsr = 0;
-    ccsr = set_field(ccsr, CCSR_ENABLE, cpu->cfg.ext_cheri);
-    ccsr = set_field(ccsr, CCSR_DIRTY, 1); /* Always report dirty */
+    ccsr = set_field(ccsr, XCCSR_ENABLE, cpu->cfg.ext_cheri);
+    ccsr = set_field(ccsr, XCCSR_DIRTY, 1); /* Always report dirty */
+
+#if !defined(TARGET_RISCV32)
+    if (csrno == CSR_SCCSR)
+        ccsr |= env->sccsr;
+#endif
+
     qemu_log_mask(CPU_LOG_INT, "Reading xCCSR(%#x): %x\n", csrno, (int)ccsr);
     *val = ccsr;
     return 0;
 }
 
 static int write_ccsr(CPURISCVState *env, int csrno, target_ulong val)
 {
-    error_report("Attempting to write " TARGET_FMT_lx
-                 "to xCCSR(%#x), this is not supported (yet?).",
-                 val, csrno);
-    return -1;
+    switch(csrno) {
+    default:
+        error_report("Attempting to write " TARGET_FMT_lx
+                     "to xCCSR(%#x), this is not supported (yet?).",
+                     val, csrno);
+        return -1;
+#if !defined(TARGET_RISCV32)
+    case CSR_SCCSR: {
+        static const target_ulong gclgmask = (SCCSR_SGCLG | SCCSR_UGCLG);
+        /* Take the GCLG bits from the store and update state bits */
+        env->sccsr = set_field(env->sccsr, gclgmask, get_field(val, gclgmask));
```

Comment:
Hmm I wonder if we need to flush the QEMU TLB after a write that changes the value here? Otherwise we might end up with stale/missing PAGE_LC_TRAP on the pages?

Reply:
Is the same thing not missing from the existing MIPS EntryHi code? It's flushing on ASID changes but not CLGx changes.